File: Mobile/lib/utils/payment_config.dart
/// Normalize legacy (`card`/`cash`) and new (`wompi`/`pay_at_venue`) payment-config API shapes.
Map<String, dynamic> normalizePaymentConfig(Map<String, dynamic>? cfg) {
  final raw = cfg ?? <String, dynamic>{};
  final defaultCommission = (raw['defaultCommission'] as num?)?.toDouble() ?? 15.0;
  final rawMethods = raw['methods'];
  final methods = <Map<String, dynamic>>[];

  if (rawMethods is List) {
    for (final item in rawMethods) {
      if (item is! Map) continue;
      final m = Map<String, dynamic>.from(item);
      final rawId = '${m['id'] ?? ''}';
      final id = _mapMethodId(rawId);
      final enabled = m['enabled'] != false;
      final configured = m['configured'] is bool
          ? m['configured'] as bool
          : _inferConfigured(rawId, raw);

      var label = '${m['label'] ?? ''}'.trim();
      if (label.isEmpty) {
        if (id == 'wompi') {
          label = 'Card';
        } else if (id == 'yappy') {
          label = 'Yappy';
        } else {
          label = 'Pay by visit';
        }
      }

      methods.add({
        'id': id,
        'label': label,
        'enabled': enabled,
        'configured': configured,
      });
    }
  }

  if (methods.isEmpty) {
    final payAtVenue =
        raw['cashPayEnabled'] != false && raw['payAtVenueEnabled'] != false;
    methods.addAll([
      {'id': 'wompi', 'label': 'Card', 'enabled': false, 'configured': false},
      {'id': 'yappy', 'label': 'Yappy', 'enabled': false, 'configured': false},
      {
        'id': 'pay_at_venue',
        'label': 'Pay by visit',
        'enabled': payAtVenue,
        'configured': true,
      },
    ]);
  }

  return {
    'defaultCommission': defaultCommission,
    'methods': methods,
  };
}

String _mapMethodId(String rawId) {
  if (rawId == 'wompi' || rawId == 'card') return 'wompi';
  if (rawId == 'yappy') return 'yappy';
  return 'pay_at_venue';
}

bool _inferConfigured(String rawId, Map<String, dynamic> cfg) {
  if (rawId == 'cash' || rawId == 'pay_at_venue') return true;
  if (rawId == 'wompi' || rawId == 'card') {
    return cfg['wompiConfigured'] == true || cfg['stripeEnabled'] == true;
  }
  if (rawId == 'yappy') return cfg['yappyConfigured'] == true;
  return false;
}

bool isPaymentMethodSelectable(Map<String, dynamic> method) {
  return method['enabled'] != false && method['configured'] == true;
}

String pickDefaultPaymentMethodId(List<Map<String, dynamic>> methods) {
  final selectable = methods.where(isPaymentMethodSelectable).toList();
  for (final m in selectable) {
    if ('${m['id']}' == 'pay_at_venue') return 'pay_at_venue';
  }
  if (selectable.isNotEmpty) return '${selectable.first['id']}';
  return 'pay_at_venue';
}

List<Map<String, dynamic>> selectablePaymentMethods(
  List<Map<String, dynamic>> methods,
) {
  return methods.where(isPaymentMethodSelectable).toList();
}

File: Web/src/lib/paymentConfig.ts
export type CheckoutPaymentId = "wompi" | "yappy" | "pay_at_venue";

export type NormalizedPaymentMethod = {
  id: CheckoutPaymentId;
  label: string;
  enabled: boolean;
  configured: boolean;
};

export type NormalizedPaymentConfig = {
  defaultCommission: number;
  methods: NormalizedPaymentMethod[];
};

function mapMethodId(id: string): CheckoutPaymentId {
  if (id === "wompi" || id === "card") return "wompi";
  if (id === "yappy") return "yappy";
  return "pay_at_venue";
}

function inferConfigured(
  rawId: string,
  cfg: Record<string, unknown>,
): boolean {
  if (rawId === "cash" || rawId === "pay_at_venue") return true;
  if (rawId === "wompi" || rawId === "card") {
    return Boolean(cfg.wompiConfigured ?? cfg.stripeEnabled);
  }
  if (rawId === "yappy") {
    return Boolean(cfg.yappyConfigured);
  }
  return false;
}

/** Normalize legacy (`card`/`cash`) and new (`wompi`/`pay_at_venue`) payment-config API shapes. */
export function normalizePublicPaymentConfig(
  cfg: Record<string, unknown> | null | undefined,
): NormalizedPaymentConfig {
  const defaultCommission =
    typeof cfg?.defaultCommission === "number" ? cfg.defaultCommission : 15;

  const rawMethods = Array.isArray(cfg?.methods) ? cfg.methods : [];
  const methods: NormalizedPaymentMethod[] = rawMethods.map((m) => {
    const row = m as Record<string, unknown>;
    const rawId = String(row.id ?? "");
    const id = mapMethodId(rawId);
    const enabled = row.enabled !== false;
    const configured =
      typeof row.configured === "boolean"
        ? row.configured
        : inferConfigured(rawId, cfg ?? {});

    let label = String(row.label ?? "").trim();
    if (!label) {
      if (id === "wompi") label = "Card";
      else if (id === "yappy") label = "Yappy";
      else label = "Pay by visit";
    }

    return { id, label, enabled, configured };
  });

  if (methods.length === 0) {
    const payAtVenue = cfg?.cashPayEnabled !== false && cfg?.payAtVenueEnabled !== false;
    return {
      defaultCommission,
      methods: [
        { id: "wompi", label: "Card", enabled: false, configured: false },
        { id: "yappy", label: "Yappy", enabled: false, configured: false },
        {
          id: "pay_at_venue",
          label: "Pay by visit",
          enabled: payAtVenue,
          configured: true,
        },
      ],
    };
  }

  return { defaultCommission, methods };
}

export function isPaymentMethodSelectable(m: NormalizedPaymentMethod): boolean {
  return m.enabled && m.configured;
}

export function pickDefaultPaymentMethod(
  methods: NormalizedPaymentMethod[],
): CheckoutPaymentId {
  const selectable = methods.filter(isPaymentMethodSelectable);
  const payAtVenue = selectable.find((m) => m.id === "pay_at_venue");
  if (payAtVenue) return "pay_at_venue";
  if (selectable.length > 0) return selectable[0].id;
  return "pay_at_venue";
}

export function selectablePaymentMethods(
  methods: NormalizedPaymentMethod[],
): NormalizedPaymentMethod[] {
  return methods.filter(isPaymentMethodSelectable);
}

/**
 * Customer-facing payment methods per the Rezervame spec (§9): card + Yappy only,
 * never "pay at venue". Backend endpoints still accept pay_at_venue — this only
 * shapes what the customer web offers.
 */
export function customerFacingPaymentMethods(
  methods: NormalizedPaymentMethod[],
): NormalizedPaymentMethod[] {
  return methods.filter((m) => m.id !== "pay_at_venue");
}

/** Default customer payment tab: prefer card, then Yappy. Never pay-at-venue. */
export function pickCustomerDefaultPaymentMethod(
  methods: NormalizedPaymentMethod[],
): Exclude<CheckoutPaymentId, "pay_at_venue"> {
  const online = customerFacingPaymentMethods(methods);
  const card = online.find((m) => m.id === "wompi");
  if (card) return "wompi";
  const yappy = online.find((m) => m.id === "yappy");
  if (yappy) return "yappy";
  return "wompi";
}
